core/src/main/scala/tapir/Codec.scala

```diff
@@ -178,41 +188,51 @@ Is there an implicit schema for: ${T}, and all of its components?
 (codecs are looked up as implicit values of type Codec[${T}, ${M}, _];
 schemas are looked up as implicit values of type SchemaFor[${T}])
 """)
-trait CodecForOptional[T, M <: MediaType, R] extends Decode[Option[R], T] { outer =>
+trait CodecForOptional[T, M <: MediaType, R] extends Validate[Option[R], T] { outer =>
   def encode(t: T): Option[R]
   def decode(s: Option[R]): DecodeResult[T]
   def meta: CodecMeta[M, R]
 
-  def mapDecode[TT](f: T => DecodeResult[TT])(g: TT => T): CodecForOptional[TT, M, R] =
+  def validate(v: Validator[T]): CodecForOptional[T, M, R] = new CodecForOptional[T, M, R] {
+    override def encode(t: T): Option[R] = outer.encode(t)
+    override def decode(s: Option[R]): DecodeResult[T] = outer.decode(s)
+    override def meta: CodecMeta[M, R] = outer.meta
+    override private[tapir] val validator: Validator[T] = v
+  }
+
+  def mapDecode[TT](f: T => DecodeResult[TT])(g: TT => T)(implicit v: Validator[TT]): CodecForOptional[TT, M, R] =
     new CodecForOptional[TT, M, R] {
       override def encode(t: TT): Option[R] = outer.encode(g(t))
       override def decode(s: Option[R]): DecodeResult[TT] = outer.decode(s).flatMap(f)
       override val meta: CodecMeta[M, R] = outer.meta
+      override private[tapir] val validator: Validator[TT] = outer.validator.contramap(g)
     }
 
-  def map[TT](f: T => TT)(g: TT => T): CodecForOptional[TT, M, R] = mapDecode[TT](f.andThen(Value.apply))(g)
+  def map[TT](f: T => TT)(g: TT => T)(implicit v: Validator[TT]): CodecForOptional[TT, M, R] = mapDecode[TT](f.andThen(Value.apply))(g)(v)
 }
 
 object CodecForOptional {
   type PlainCodecForOptional[T] = CodecForOptional[T, MediaType.TextPlain, String]
 
-  implicit def fromCodec[T, M <: MediaType, R](implicit c: Codec[T, M, R]): CodecForOptional[T, M, R] = new CodecForOptional[T, M, R] {
-    override def encode(t: T): Option[R] = Some(c.encode(t))
-    override def decode(s: Option[R]): DecodeResult[T] = s match {
-      case None    => DecodeResult.Missing
-      case Some(h) => c.decode(h)
+  implicit def fromCodec[T, M <: MediaType, R](implicit c: Codec[T, M, R]): CodecForOptional[T, M, R] =
+    new CodecForOptional[T, M, R] {
+      override def encode(t: T): Option[R] = Some(c.encode(t))
+      override def decode(s: Option[R]): DecodeResult[T] = s match {
+        case None    => DecodeResult.Missing
+        case Some(h) => c.decode(h)
+      }
+      override val meta: CodecMeta[M, R] = c.meta
```

Comment:
the validator should be inherited from the codec, no?

Reply:
Now it is, because I moved it into `CodecMeta`